### ISSUE TITLE
Failure to create concrete instance of entity navigation collection class

### ISSUE THREAD

**Issue by bradmarder** [aspnet/EntityFrameworkCore#11923]
This error message appears to be masking some other internal error. Even if I change all collections to `List<>`, it has no effect. 

```
Unhandled Exception: System.InvalidOperationException: The type of navigation property 'Paragraphs' on the entity type 'Post' is 'ICollection<Paragraph>' for which it was not possible to create a concrete instance. Either initialize the property before use, add a public parameterless constructor to the type, or use a type which can be assigned a HashSet<> or List<>.
   at Microsoft.EntityFrameworkCore.Metadata.Internal.ClrICollectionAccessor`3.Create()
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.ProjectionExpressionVisitor.MaterializeCollectionNavigationAsync[TEntity](INavigation navigation, IAsyncEnumerable`1 elements, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.TaskLiftingExpressionVisitor._ExecuteAsync[T](IReadOnlyList`1 taskFactories, Func`2 selector)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.AsyncSelectEnumerable`2.AsyncSelectEnumerator.MoveNext(CancellationToken cancellationToken)
   at System.Linq.AsyncEnumerable.First_[TSource](IAsyncEnumerable`1 source, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.TaskLiftingExpressionVisitor._ExecuteAsync[T](IReadOnlyList`1 taskFactories, Func`2 selector)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.AsyncSelectEnumerable`2.AsyncSelectEnumerator.MoveNext(CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncLinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext(CancellationToken cancellationToken)
   at System.Linq.AsyncEnumerable.Aggregate_[TSource,TAccumulate,TResult](IAsyncEnumerable`1 source, TAccumulate seed, Func`3 accumulator, Func`2 resultSelector, CancellationToken cancellationToken)
   at EFCoreBugs.Program.Main(String[] args) in C:\Users\Brad\source\repos\ConcreteBug\ConcreteBug\Program.cs:line 34
   at EFCoreBugs.Program.<Main>(String[] args)
```

```c#
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Threading.Tasks;

using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Console;

namespace EFCoreBugs
{
    class Program
    {
        static async Task Main(string[] args)
        {
            using (var db = new TestDb())
            {
                await db.Database.EnsureDeletedAsync();
                await db.Database.EnsureCreatedAsync();

                var Fblog = new Blog();
                var Fpost = new Post();
                var FPara = new Paragraph();
                
                Fpost.Paragraphs.Add(FPara);
                Fblog.Posts.Add(Fpost);
                db.Blogs.Add(Fblog);

                await db.SaveChangesAsync();
            }
            using (var db = new TestDb())
            {
                await db.Blogs
                    .Select(blog => new
                    {
                        Foo = blog.Posts.First().Paragraphs.Sum(x => 5)
                    })
                    .ToListAsync();
            }
            
            Console.ReadLine();
        }
    }

    public class TestDb : DbContext
    {
        public DbSet<Blog> Blogs { get; set; }
        public DbSet<Post> Posts { get; set; }
        public DbSet<Paragraph> Paragraphs { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            var log = new LoggerFactory(new[] { new ConsoleLoggerProvider((_, __) => true, true) });
            optionsBuilder.UseLoggerFactory(log);
            optionsBuilder.UseSqlServer("Data Source=(localdb)\\MSSQLLocalDB;Initial Catalog=master;Database=bugtest;Integrated Security=True;Connect Timeout=30;Encrypt=False;TrustServerCertificate=True;ApplicationIntent=ReadWrite;MultiSubnetFailover=False");
        }
    }

    public class Blog
    {
        public Guid Id { get; set; }

        [InverseProperty(nameof(Post.Blog))]
        public virtual ICollection<Post> Posts { get; } = new HashSet<Post>();
    }

    public class Post
    {
        public Guid Id { get; set; }
        public Guid BlogId { get; set; }

        [ForeignKey(nameof(BlogId))]
        public virtual Blog Blog { get; private set; }

        [InverseProperty(nameof(Paragraph.Post))]
        public virtual ICollection<Paragraph> Paragraphs { get; } = new HashSet<Paragraph>();
    }

    public class Paragraph
    {
        public Guid Id { get; set; }
        public Guid PostId { get; set; }

        [ForeignKey(nameof(PostId))]
        public virtual Post Post { get; private set; }
    }
}

```

### Further technical details
EF Core version: 2.1.0-preview2-final
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: win10 x64
IDE:  Visual Studio 2017 15.7.0 Preview 6.0

**Comment by maumar** [comment]
@bradmarder you need to add setters to your navigation properties, e.g.:

```
[InverseProperty(nameof(Paragraph.Post))]
        public virtual ICollection<Paragraph> Paragraphs { get; set; } = new HashSet<Paragraph>();
```

**Comment by maumar** [comment]
fixed in 206a3bf164659f345362cbb1e2e2e30ffa9edb1b
